DIPs/1NNN-AP.md

```diff
@@ -0,0 +1,172 @@
+# Enum Type Inference
+
+| Field           | Value                                                           |
+|-----------------|-----------------------------------------------------------------|
+| DIP:            |                                                                 |
+| Review Count:   | 0                                                               |
+| Author:         | Aya Partridge (zxinsworld@gmail.com)                            |
+| Implementation: |                                                                 |
+| Status:         | Draft                                                           |
+
+## Abstract
+
+Enum Type Inference (ETI) is a shortcut to allow the omission of an enum member's
+type name when it can be contextually inferred.
+
+While D's pre-existing syntax makes ETI a more complex feature to implement than in brand
+new languages, its presence in many new languages has proven its convenience and popularity.
+
+## Contents
+* [Rationale](#rationale)
+* [Prior Work](#prior-work)
+* [Description](#description)
+* [Reference](#reference)
+* [Copyright & License](#copyright--license)
+* [Reviews](#reviews)
+
+## Rationale
+Writing the same enum type names over and over again (e.g. exhaustive enum switch-cases)
+involves the unnecessary, often gratuitous repetition of the enum member's type name.
+
+The solution used by many other modern languages is simple: permit the omission
+of the enum member's type name when it can be inferred from its context.
+Dlang having this shortcut would be equally beneficial, with few drawbacks.
+
+## Prior Work
+Implementation of this feature in other languages:
+- [Swift](https://docs.swift.org/swift-book/LanguageGuide/Enumerations.html)
+- [Ziglang](https://ziglang.org/documentation/master/#Enum-Literals)
+- [Odin](https://odin-lang.org/docs/overview/#implicit-selector-expression)
+
+Java allows omitting an enum type in switch-case statements: [https://docs.oracle.com/javase/tutorial/java/javaOO/enum.html](https://docs.oracle.com/javase/tutorial/java/javaOO/enum.html)
+
+A suggestion for a similar feature in Rust, but with a different approach: [https://internals.rust-lang.org/t/enum-path-inference-with-variant-syntax/16851](https://internals.rust-lang.org/t/enum-path-inference-with-variant-syntax/16851)
+
+## Description
+Due to `.enumMember` being the same syntax as the [module scope operator](https://dlang.org/spec/module.html#module_scope_operators),
+I propose that ETI should use the syntax `$enumMember`. I'm open to discussing alternatives,
+but I believe that my proposed syntax represents the best compromise between convenience and compatibility.
+
+#### 1. Assignment statements.
+Assigning to a variable with a known type should allow ETI.
+```d
+enum A{ a,b,c,d; }
+
+struct B{ A one, two; }
+
+void main(){
+    A    myA1 = $b; //myA1 = A.b
```

Comment:
Should compound expressions work? E.g. `A x = $a | $b;`. Ditto for `return $a | $b;`.

Reply:
> Should compound expressions work? E.g. `A x = $a | $b;`. Ditto for `return $a | $b;`.

I think I'm unfamiliar with this syntax, what does bitwise ORing two enum members result in?

Reply:
The same thing as if you were bitwise ORing their base type.

Reply:
Thanks. Just to point out, enum ORing is typically used with non-overlapping enum values such as a=1, b=2, c=4, d=8, e=16 etc to pass multiple flags as a function argument. So a | c would be 5. Then the parameter received can be ANDed with c to see if c is set.

Reply:
> Thanks. Just to point out, enum ORing is typically used with non-overlapping enum values such as a=1, b=2, c=4, d=8, e=16 etc to pass multiple flags as a function argument. So a | c would be 5. Then the parameter received can be ANDed with c to see if c is set.

Saying that, I think a syntax like this would be very useful for C API wrappers. would Do you know if it would be feasible to make this work without breaking changes?
```d
enum Flags{ a=1,b=2,c=4,d=8 }
void setFlags(int flags);
int flagVar;

void main(){
    setFlags(Flags.a | $b | $d);
    flagVar = Flags.d + $a + $a;
    flagVar = Flags.b & $c;
}
```
I'm picturing it having the same rules as the array literal syntax. (first—or maybe closest to the left?—enum type used for ETI, can be interchanged with integers, maybe floats)